egui/src/input_state/touch_state.rs

```diff
@@ -0,0 +1,288 @@
+    /// `MultiTouchInfo` is created.
+    pub num_touches: usize,
+    /// Zoom factor (Pinch or Zoom).  Moving fingers closer together or further appart will change
+    /// this value.  This is a relative value, comparing the average distances of the fingers in
+    /// the current and previous frame.  If the fingers did not move since the previous frame,
+    /// this value is `1.0`.
+    pub zoom_delta: f32,
+    /// Rotation in radians.  Moving fingers around each other will change this value.  This is a
+    /// relative value, comparing the orientation of fingers in the current frame with the previous
+    /// frame.  If all fingers are resting, this value is `0.0`.
+    pub rotation_delta: f32,
+    /// Relative movement (comparing previous frame and current frame) of the average position of
+    /// all touch points.  Without movement this value is `Vec2::ZERO`.
+    ///
+    /// Note that this may not necessarily be measured in screen points (although it _will_ be for
+    /// most mobile devices).  In general (depending on the touch device), touch coordinates cannot
+    /// be directly mapped to the screen.  A touch always is considered to start at the position of
+    /// the pointer, but touch movement is always measured in the units delivered by the device,
+    /// and may depend on hardware and system settings.
+    pub translation_delta: Vec2,
+    /// Current force of the touch (average of the forces of the individual fingers). This is a
+    /// value in the interval `[0.0 .. =1.0]`.
+    ///
+    /// Note 1: A value of `0.0` either indicates a very light touch, or it means that the device
+    /// is not capable of measuring the touch force at all.
+    ///
+    /// Note 2: Just increasing the physical pressure without actually moving the finger may not
+    /// necessarily lead to a change of this value.
+    pub force: f32,
+}
+
+/// The current state (for a specific touch device) of touch events and gestures.
+#[derive(Clone)]
+pub(crate) struct TouchState {
+    /// Technical identifier of the touch device.  This is used to identify relevant touch events
+    /// for this `TouchState` instance.
+    device_id: TouchDeviceId,
+    /// Active touches, if any.
+    ///
+    /// TouchId is the unique identifier of the touch.  It is valid as long as the finger/pen touches the surface.  The
+    /// next touch will receive a new unique ID.
+    ///
+    /// Refer to [`ActiveTouch`].
+    active_touches: BTreeMap<TouchId, ActiveTouch>,
+    /// If a gesture has been recognized (i.e. when exactly two fingers touch the surface), this
+    /// holds state information
+    gesture_state: Option<GestureState>,
+}
+
+#[derive(Clone, Debug)]
+struct GestureState {
+    start_time: f64,
+    start_pointer_pos: Pos2,
+    force: f32,
+    previous: Option<DynGestureState>,
+    current: DynGestureState,
+}
+
+/// Gesture data which can change over time
+#[derive(Clone, Copy, Debug)]
+struct DynGestureState {
```

Comment:
Could perhaps be combined with `TouchAverages` for some shorter code

Reply:
Resolved with <https://github.com/emilk/egui/pull/306/commits/3918653979261912606a2a6d184c0c964fdcc1ce>. I removed `TouchAverages` completely and moved the `force` field from `GestureState` to `DynGestureState`.